algebird-test/src/test/scala/com/twitter/algebird/AggregatorLaws.scala

```diff
@@ -89,7 +89,11 @@ class AggregatorLaws extends CheckProperties {
   def checkNumericSum[T: Arbitrary](implicit num: Numeric[T]): Prop =
     forAll { in: List[T] =>
       val aggregator = Aggregator.numericSum[T]
-      aggregator(in) == in.map(num.toDouble).sum
```

Comment:
I added a few extra tests and I think I've uncovered... what MIGHT be a bigger problem here that I'll submit to your enormous pulsing brain for consideration.

Here are a few tests that check:

- that `input.sum` accumulates errors, as we expect
- that Kahan summation is able to control these errors

I did this by accumulating 1000 `0.01`s.

```
import org.scalatest.funsuite.AnyFunSuite

/**
  Unit tests to highlight specific examples of the properties we guarantee.
  */
class AggregatorTests extends AnyFunSuite {
  test("Kahan summation mitigates error accumulation in Double streams") {

    val input = Stream.continually(0.01).take(1000)

    assert(9.999999999999831 == input.sum, "naive summation accumulates errors")
    assert(Some(10.0) == DoubleRing.sumOption(input), "Kahan summation controls error accumulation")

    val aggregator = Aggregator.numericSum[Double]
    assert(10.0 == aggregator(input), "aggregators control error accumulation")
  }

  test("Kahan summation mitigates error accumulation in Float streams") {
    val input = Stream.continually(0.01f).take(1000)

    assert(10.0001335f == input.sum, "naive summation accumulates errors")
    assert(Some(10.0f) == FloatRing.sumOption(input), "Kahan summation controls error accumulation")

    val aggregator = Aggregator.numericSum[Float]
    assert(10.0f == aggregator(input), "agg control error accumulation")
  }
}
```

Here are the results:

```
[info] AggregatorTests:
[info] - Kahan summation mitigates error accumulation in Double streams
[info] - Kahan summation mitigates error accumulation in Float streams *** FAILED ***
[info]   10.0 did not equal 9.999999776482582 agg control error accumulation (AggregatorLaws.scala:46)
```

`FloatRing.sumOption` and `DoubleRing.sumOption` are both doing the right thing here, but `Aggregator.numericSum[Float]` is NOT doing the right thing, and accumulating its own errors such that the sum == `9.999999776482582` .

Reply:
Ah, here's the problem:

```scala
scala> val input = Stream.continually(0.01f).take(1000).map(implicitly[Numeric[Float]].toDouble(_))
input: scala.collection.immutable.Stream[Double] = Stream(0.009999999776482582, ?)

scala> input.map(implicitly[Numeric[Float]].toDouble(_)).take(5).toList
res3: List[Double] = List(0.009999999776482582, 0.009999999776482582, 0.009999999776482582, 0.009999999776482582, 0.009999999776482582)
```

So the double conversion introduces trouble here.

```scala
def numericSum[T](implicit num: Numeric[T]): MonoidAggregator[T, Double, Double] =
    Preparer[T].map(num.toDouble).monoidAggregate(Aggregator.fromMonoid)
```

Here's a better set of tests from `Aggregator.fromMonoid` that pass:

```scala
import org.scalatest.funsuite.AnyFunSuite

/**
  Unit tests to highlight specific examples of the properties we guarantee.
  */
class AggregatorTests extends AnyFunSuite {
  test("Kahan summation mitigates Double error accumulation") {
    val input = Stream.continually(0.01).take(1000)

    assert(9.999999999999831 == input.sum, "naive summation accumulates errors")
    assert(Some(10.0) == DoubleRing.sumOption(input), "Kahan summation controls error accumulation")
    assert(10.0 == Aggregator.numericSum[Double].apply(input))
  }

  test("Kahan summation mitigates Float error accumulation") {
    val input = Stream.continually(0.01f).take(1000)

    assert(10.0001335f == input.sum, "naive summation accumulates errors")
    assert(Some(10.0f) == FloatRing.sumOption(input), "Kahan summation controls error accumulation")
    assert(10.0f == Aggregator.fromMonoid[Float].apply(input))
  }
}
```

BTW, with `import Ring.numericRing` the float tests fail, since `NumericRing[Float]` doesn't get access to the Kahan summation trick.

The docs advise

```scala
* Note that if you instead wanted to aggregate Numeric values of a type T into the same type T
   * (e.g. if you want MonoidAggregator[T, T, T] for some Numeric type T), you can directly use
   * Aggregator.fromMonoid[T] after importing the numericRing implicit:
   *
   *   > import com.twitter.algebird.Ring.numericRing
   *   > def numericAggregator[T: Numeric]: MonoidAggregator[T, T, T] = Aggregator.fromMonoid[T]
```

Which is maybe not the correct advice.